walkthrough-w-4 (GCP API's) Sentiment Analysis and Google Maps API integrations in portfolio

dmosc opened the pull request:
Week 4 walk-through integration in portfolio.

+ Added Google Maps and GCP Sentiment Analysis API integrations to support sentiment analysis on submitted messages by users and a Map to display places I go to frequently at the end of the portfolio.

*** No breaking changes ***

dmosc committed:
feature (Sentiment Analysis) Integrate sentiment analysis API

+ Sentiment Analysis GCP API is used to add sentiment to submitted comments
by users. This feature tags and categorizes each submitted comment.

*** No breaking changes ***

dmosc committed:
feature (Google Maps) Integrate Maps API with portfolio.

+ Added a map at the end of the portfolio with marked places that
I go to a lot. The div with id "map" holds the generated Map inside the
queries/index.js file.

*** No breaking changes ***

anniefu commented on portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java:
```diff
@@ -126,4 +131,22 @@ private String getBody(HttpServletRequest request) throws IOException {
         return "";
     }
+
+    private double getSentiment(String comment) throws IOException {
```
[nit] I would rename this to something like "analyzeSentiment", the name collision with the Comment class' getSentiment was a bit confusing.

anniefu commented on portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java:
```diff
@@ -126,4 +131,22 @@ private String getBody(HttpServletRequest request) throws IOException {
+            .setType(Document.Type.PLAIN_TEXT)
+            .build();
+
+        LanguageServiceClient languageService = LanguageServiceClient.create();
```
I'd make this a private variable on this class so you can re-use the language service on every POST instead of re-creating it. 

In this particular case, there's probably not much overhead to calling LanguageServiceClient.create(), but you could imagine if this was something like connect to database, it might be wasteful to re-connect every time.

dmosc replied:
Yeah that's true. But I am not sure if it'd be possible to instantiate the service once, because as I've read, every time to finish using it, you have to close the service. But if it gets closed after the first request and then not instantiated again, it will not be available.

I am thinking probably of opening the service on the constructor, and closing it on the destructor to reduce that overhead that you're mentioning it. WDYT?

anniefu replied:
Hmm, I see what you mean about needing to close. Ok, I see two equally reasonable options here.

1: Just ignore my suggestion, and just use the try-with-resources block style they have in the sample code to make it a little cleaner: https://cloud.google.com/natural-language/docs/reference/libraries#client-libraries-usage-java

2: Implement init/destroy of HttpServlet Class https://docs.oracle.com/javaee/7/api/javax/servlet/http/HttpServlet.html and calling close() in the destroy. which is pretty much the same as your constructor/destructor suggestion but specific to HttpServlet.


Which one is better depends on how many comments you expect to get. If comments are very frequent, it's probably worth keeping a connection to the service around. If comments are fewer or more spread out, it's probably not that big of deal to just reconnect for each comment.

For a portfolio site, you can reasonably assume fewer comments, but maybe a site like Reddit would expect frequent enough comments to prefer keeping service connections alive.

dmosc replied:
Yeah, sounds reasonable. I guess the amount of comments submitted does really impact the implementation. I think I will stick to the try-with-resources approach to clean up the code since it is true that for a webapp of this type we shouldn't be expecting massive interaction.

Thanks for the options though. It's nice to know the different approaches.

anniefu commented on portfolio/src/main/webapp/functions/index.js:
```diff
@@ -1,11 +1,26 @@
+        "Happy";
+
+    const badgeToSet =
+        score >= -1 && score <= -0.3 ? "danger" :
```
These look to be the same ranges as the "emotion". If that's intentional, consider using the emotion to set the badge instead. That way if you change the ranges in the future, you don't have to do it in both places.

If you think in the future, you might want the emotion and badge ranges to be different, then it's perfectly fine to leave it as-is.

anniefu commented on portfolio/src/main/webapp/functions/index.js:
```diff
@@ -1,11 +1,26 @@
+
+    const score = parseFloat(commentPayload.sentiment);
+    const emotion =
+        score >= -1 && score <= -0.3 ? "Mad" :
```
[super nit] how come these ones are capitalized, but the badges aren't?

dmosc replied:
These ones are the actual text added to the HTML element, and the other ones are Bootstrap CSS classes. So their class naming is all lowercase, and the values "Happy", "Mad", "Neutral", are the textual value displayed.

anniefu replied:
Ah, ok. Thanks for explaining!

anniefu commented on portfolio/src/main/webapp/requests/queries/index.js:
```diff
@@ -78,6 +240,52 @@ const loadMoreProjects =  async () => {
+        styles: [...mapStyles]
+    });
+
+    const places = [
```
Nice! Just a thing to be cautious about, I don't know the geographic spread of these locations well, but I'd be careful to avoid putting too many near your home address. You probably want to avoid me being able to geolocate your home address within a few miles based off your favorite places.

If they're all near your university, then that should be fine since there's tons of students in that area.

dmosc replied:
Yeah I guess I am exposing myself right hahahah.

But yeah, everything is around the university.

ssmall commented on portfolio/src/main/webapp/functions/index.js:
```diff
@@ -1,11 +1,26 @@
     comment.innerHTML = commentPayload.comment;
+
+    const score = parseFloat(commentPayload.sentiment);
+    const emotion =
+        score >= -1 && score <= -0.3 ? "Mad" :
+        score > -0.3 && score <= 0.3 ? "Neutral" :
+        "Happy";
```
Consider moving the logic to translate the score to a string to the backend. That way, the frontend only needs to worry about display logic, not business logic.

ssmall commented on portfolio/src/main/webapp/requests/queries/index.js:
```diff
@@ -78,6 +240,52 @@ const loadMoreProjects =  async () => {
     projects.scrollTo({ top: projects.scrollHeight, behavior: "smooth" });
 };
 
+const loadMap = () => {
```
nit: the name of this file is `queries/index.js`, but this function is not related to queries. Consider moving the map functionality to a different file.

dmosc replied:
What would be a nice approach?

It's cause, normally, I'd use the "queries" folder as a suite of functions to fetch any type of info (a server endpoint or an API endpoint). But I would agree that technically they are not the same.

I could probably create a folder for API requests that don't go to the webapp's server. WDYT?

ssmall replied:
I would recommend splitting out the files by which different services they interact with. In that case, it would make sense to move the functions related to maps to a separate directory named something like `maps/index.js`.

anniefu commented on portfolio/src/main/webapp/functions/index.js:
```diff
@@ -18,7 +33,13 @@ const createCommentElement = commentPayload => {
+    tagsSection.classList.add("header-tags");
+
+    tagsSection.appendChild(username);
+    tagsSection.appendChild(sentiment);
```
I think it's pretty nifty that you're displaying your sentiment analysis usage. Though it might be interesting to have a sentiment summary page somewhere where you can see sentiment stats, but without the usernames/comments attached, instead of displaying it directly on the comment.

If user's can see their comments and the analysis, it might influence what comment they write before they even write it.

Not a change you need to do in this PR, though, and it's totally cool to leave it as is. The main purpose of the portfolio is to let you showcase your dev skills, so that's not lost here.

dmosc replied:
Yeah, probably a separate dashboard with charts to visualize the sentiment of comments would be a great option to display the results of the comments.

ssmall reviewed (changes requested):
Looks good, Oscar! Can you please give this PR a more descriptive title, though?

dmosc committed:
nitfix (Resolve comments) Resolve nitfixes and code refactoring.

+ Add {emotion: String} attribute to Comment.java to store a descriptive
tag of the comment's sentiment.

+ Modified CommentsServlet.java to handle the new Comment.java emotion attribute
on querying and creating comments. The tag for new comments is defined in doPost.
Changed getSentiment() to analyzeSentiment() to avoid conflicting with getters
and setters naming convention.

+ Removed emotion tag creation in functions/index.js since tag is now set
on the backend service. This way the webapp remains agnostic to the business
logic.

+ Add new google-maps/index.js import in index.html.

+ Refactored the loadMap() function from queries/index.js to a new folder
google-maps/index.js to separate and group types of queries involving
separate API's.

*** No breaking changes ***

ssmall commented on portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java:
```diff
@@ -80,21 +85,27 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
         String comment = getAttribute(json, "comment", "...");
+        long timestamp = System.currentTimeMillis();
+        double sentiment = analyzeSentiment(comment);
+        String emotion =
+            sentiment >= -1 && sentiment <= -0.3 ? "Mad" :
+            sentiment > -0.3 && sentiment <= 0.3 ? "Neutral" :
+            "Happy";
```
Nit: it seems like it might be beneficial to use an `Enum` for sentiment. Just something to consider, but the code is also totally okay as-is.

ssmall commented on portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java:
```diff
@@ -126,4 +137,23 @@ private String getBody(HttpServletRequest request) throws IOException {
+                .getDocumentSentiment()
+                .getScore();
+
+            languageService.close();
```
Because you added the try-with-resources on line 142, it is no longer necessary to call `languageService.close()`. In Java, classes that implement [AutoCloseable](https://docs.oracle.com/javase/7/docs/api/java/lang/AutoCloseable.html?is-external=true) (which LanguageServiceClient [does](https://googleapis.dev/java/google-cloud-language/latest/com/google/cloud/language/v1/LanguageServiceClient.html)) will have their `close()` methods called automatically upon exiting the `try` block.

ssmall commented on portfolio/src/main/java/com/google/sps/classes/Comment.java:
```diff
@@ -25,6 +29,14 @@ public long getTimeStamp() {
         return timestamp;
     }
 
+    public double getSentiment() {
+        return sentiment;
+    }
```
Is this value used anywhere on the frontend? If not, it can probably be removed.

dmosc replied:
I was thinking that maybe, if in the future I'd like to create an admin dashboard to display the analytics and the results of the comments, it would be better to store the numeric value of the sentiment rather than the sole tag.

ssmall replied:
That makes sense, but there's a common mantra in agile software development: [YAGNI - You Aren't Gonna Need It](https://www.martinfowler.com/bliki/Yagni.html). Basically the idea is to defer adding fields/methods/features until you have a real use for them. It this case, adding a new field to the `Comment` class is a backwards-compatible change, so you can add it at any point in the future with little risk of breaking anything.

dmosc replied:
Nice mantra!
Well, I'll go ahead and remove it.

ssmall reviewed (commented):
This is looking really good! I have just a few more small comments.

dmosc committed:
nitfix (Enum) Change harcoded emotion strings for Enum.

+ Removed harcoded string emotions in CommentsServlet.java doPost
to group them in an Enum instead.

*** No breaking changes ***

ssmall commented on portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java:
```diff
@@ -80,21 +85,26 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
         String comment = getAttribute(json, "comment", "...");
+        long timestamp = System.currentTimeMillis();
+        double sentiment = analyzeSentiment(comment);
+        Emotion emotion =
+            sentiment >= -1 && sentiment <= -0.3 ? Emotion.Mad :
+            sentiment > -0.3 && sentiment <= 0.3 ? Emotion.Neutral :
+            Emotion.Happy;
```
this is an FYI comment, so please don't worry about changing the code. I'm just including it for educational purposes.

The idiomatic Java way to implement this would be to add a `static` method to the `Emotion` enum, something like:
```java
private enum Emotion {
    Happy,
    Neutral,
    Mad;

    static Emotion fromSentiment(double sentiment) {
        return sentiment >= -1 && sentiment <= -0.3 ? Mad :
               sentiment > -0.3 && sentiment <= 0.3 ? Neutral :
               Happy;
    }
}
```

and then this line would just become

```java
Emotion emotion = Emotion.fromSentiment(sentiment);
```

dmosc replied:
Oh that is totally better!
I didn't know you could add methods to enums. Will definitely look into that!

ssmall reviewed (approved):
```
                __/___            
          _____/______|           
  _______/_____\_______\_____     
  \              < < <       |    
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Ship it!
